### PR TITLE
Fix: replace most of bare pointers with weak/unique_ptr in module_lr to prevent memory leak (Useful information about how to use smart pointers in C++11))

### LINKED ISSUE #4718: Bug: memory leak detected in LR-TDDFT integrated test
### Describe the bug

see https://[raw.githubusercontent.com/deepmodeling/abacus-develop/dashboard/README.md](https://raw.githubusercontent.com/deepmodeling/abacus-develop/dashboard/README.md)

ctrl+F then search "abacus". Seems there are plenty of tests were detected to have memory leak issue, I just copy two of them:
### On process id asan.113268

```bash

=================================================================
==113268==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 4816 byte(s) in 7 object(s) allocated from:
    #0 0x7fd503b4b887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7fd502d7abc6 in ompi_op_base_op_select (/lib/x86_64-linux-gnu/libmpi.so.40+0xd6bc6)

Direct leak of 4816 byte(s) in 7 object(s) allocated from:
    #0 0x7fd503b4b887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7fd4c11f04fb  (<unknown module>)

Direct leak of 2760 byte(s) in 4 object(s) allocated from:
    #0 0x7fd503b4ba57 in __interceptor_calloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:154
    #1 0x7fd4c31a8c13  (<unknown module>)
    #2 0x3537  (<unknown module>)

Direct leak of 1080 byte(s) in 1 object(s) allocated from:
    #0 0x7fd503b4d1e7 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x55fefe015e07 in ModuleESolver::ESolver_FP::ESolver_FP() /__w/abacus-develop/abacus-develop/source/module_esolver/esolver_fp.cpp:13
    #2 0x55fefe76981c in LR::ESolver_LR<std::complex<double>, double>::ESolver_LR(ModuleESolver::ESolver_KS_LCAO<std::complex<double>, double>&&, Input_para const&, UnitCell&) /__w/abacus-develop/abacus-develop/source/module_lr/esolver_lrtd_lcao.cpp:115
    #3 0x55fefdff0968 in ModuleESolver::init_esolver(Input_para const&, UnitCell&) /__w/abacus-develop/abacus-develop/source/module_esolver/esolver.cpp:230
    #4 0x55fefdbb2845 in Driver::driver_run() /__w/abacus-develop/abacus-develop/source/driver_run.cpp:49
    #5 0x55fefdbac3de in Driver::atomic_world() /__w/abacus-develop/abacus-develop/source/driver.cpp:188
    #6 0x55fefdbb15d6 in Driver::init() /__w/abacus-develop/abacus-develop/source/driver.cpp:41
    #7 0x55fefd25a18a in main /__w/abacus-develop/abacus-develop/source/main.cpp:42

Direct leak of 260 byte(s) in 1 object(s) allocated from:
    #0 0x7fd503b4b887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7fd4c31f0226  (<unknown module>)

Direct leak of 256 byte(s) in 2 object(s) allocated from:
    #0 0x7fd503b4b887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7fd4c32080f1  (<unknown module>)

Direct leak of 64 byte(s) in 1 object(s) allocated from:
    #0 0x7fd503b4b887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7fd4c328c323  (<unknown module>)

Direct leak of 48 byte(s) in 1 object(s) allocated from:
    #0 0x7fd503b4b887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7fd4c321483d  (<unknown module>)

Direct leak of 46 byte(s) in 2 object(s) allocated from:
    #0 0x7fd503b4b887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7fd4e8b5f157  (/lib/x86_64-linux-gnu/libc.so.6+0x88157)

Direct leak of 32 byte(s) in 1 object(s) allocated from:
    #0 0x7fd503b4d1e7 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x55fefe70e325 in LR::PotHxcLR::PotHxcLR(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, ModulePW::PW_Basis const*, UnitCell const*, Charge const*) /__w/abacus-develop/abacus-develop/source/module_lr/potentials/pot_hxc_lrtd.cpp:24
    #2 0x55fefe7487e7 in LR::ESolver_LR<std::complex<double>, double>::init_pot(Charge const&) /__w/abacus-develop/abacus-develop/source/module_lr/esolver_lrtd_lcao.cpp:502
    #3 0x55fefe76b4de in LR::ESolver_LR<std::complex<double>, double>::ESolver_LR(ModuleESolver::ESolver_KS_LCAO<std::complex<double>, double>&&, Input_para const&, UnitCell&) /__w/abacus-develop/abacus-develop/source/module_lr/esolver_lrtd_lcao.cpp:187
    #4 0x55fefdff0968 in ModuleESolver::init_esolver(Input_para const&, UnitCell&) /__w/abacus-develop/abacus-develop/source/module_esolver/esolver.cpp:230
    #5 0x55fefdbb2845 in Driver::driver_run() /__w/abacus-develop/abacus-develop/source/driver_run.cpp:49
    #6 0x55fefdbac3de in Driver::atomic_world() /__w/abacus-develop/abacus-develop/source/driver.cpp:188
    #7 0x55fefdbb15d6 in Driver::init() /__w/abacus-develop/abacus-develop/source/driver.cpp:41
    #8 0x55fefd25a18a in main /__w/abacus-develop/abacus-develop/source/main.cpp:42

Direct leak of 30 byte(s) in 1 object(s) allocated from:
    #0 0x7fd503b4b887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7fd4c0bb74e6  (<unknown module>)

Direct leak of 23 byte(s) in 1 object(s) allocated from:
    #0 0x7fd503af29a7 in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cpp:454
    #1 0x7fd4c3192f23  (<unknown module>)

Direct leak of 20 byte(s) in 1 object(s) allocated from:
    #0 0x7fd503af29a7 in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cpp:454
    #1 0x7fd4c0af6570  (<unknown module>)
    #2 0x3331312e32396334  (<unknown module>)

Direct leak of 5 byte(s) in 1 object(s) allocated from:
    #0 0x7fd503af29a7 in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cpp:454
    #1 0x7fd4c3192d65  (<unknown module>)

Direct leak of 1 byte(s) in 1 object(s) allocated from:
    #0 0x7fd503af29a7 in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cpp:454
    #1 0x7fd4c0beb0f4  (<unknown module>)

SUMMARY: AddressSanitizer: 14257 byte(s) leaked in 32 allocation(s).
```

### On process id asan.113269

```bash

=================================================================
==113269==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 4816 byte(s) in 7 object(s) allocated from:
    #0 0x7f766c380887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7f7629a254fb  (<unknown module>)

Direct leak of 4816 byte(s) in 7 object(s) allocated from:
    #0 0x7f766c380887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7f766b5afbc6 in ompi_op_base_op_select (/lib/x86_64-linux-gnu/libmpi.so.40+0xd6bc6)

Direct leak of 2760 byte(s) in 4 object(s) allocated from:
    #0 0x7f766c380a57 in __interceptor_calloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:154
    #1 0x7f762b9cdc13  (<unknown module>)
    #2 0x3537  (<unknown module>)

Direct leak of 1080 byte(s) in 1 object(s) allocated from:
    #0 0x7f766c3821e7 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x5624228b9e07 in ModuleESolver::ESolver_FP::ESolver_FP() /__w/abacus-develop/abacus-develop/source/module_esolver/esolver_fp.cpp:13
    #2 0x56242300d81c in LR::ESolver_LR<std::complex<double>, double>::ESolver_LR(ModuleESolver::ESolver_KS_LCAO<std::complex<double>, double>&&, Input_para const&, UnitCell&) /__w/abacus-develop/abacus-develop/source/module_lr/esolver_lrtd_lcao.cpp:115
    #3 0x562422894968 in ModuleESolver::init_esolver(Input_para const&, UnitCell&) /__w/abacus-develop/abacus-develop/source/module_esolver/esolver.cpp:230
    #4 0x562422456845 in Driver::driver_run() /__w/abacus-develop/abacus-develop/source/driver_run.cpp:49
    #5 0x5624224503de in Driver::atomic_world() /__w/abacus-develop/abacus-develop/source/driver.cpp:188
    #6 0x5624224555d6 in Driver::init() /__w/abacus-develop/abacus-develop/source/driver.cpp:41
    #7 0x562421afe18a in main /__w/abacus-develop/abacus-develop/source/main.cpp:42

Direct leak of 260 byte(s) in 1 object(s) allocated from:
    #0 0x7f766c380887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7f762ba15226  (<unknown module>)

Direct leak of 256 byte(s) in 2 object(s) allocated from:
    #0 0x7f766c380887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7f762ba2d0f1  (<unknown module>)

Direct leak of 64 byte(s) in 1 object(s) allocated from:
    #0 0x7f766c380887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7f762bab1323  (<unknown module>)

Direct leak of 59 byte(s) in 3 object(s) allocated from:
    #0 0x7f766c380887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7f7651395157  (/lib/x86_64-linux-gnu/libc.so.6+0x88157)

Direct leak of 48 byte(s) in 1 object(s) allocated from:
    #0 0x7f766c380887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7f762ba3983d  (<unknown module>)

Direct leak of 32 byte(s) in 1 object(s) allocated from:
    #0 0x7f766c3821e7 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x562422fb2325 in LR::PotHxcLR::PotHxcLR(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, ModulePW::PW_Basis const*, UnitCell const*, Charge const*) /__w/abacus-develop/abacus-develop/source/module_lr/potentials/pot_hxc_lrtd.cpp:24
    #2 0x562422fec7e7 in LR::ESolver_LR<std::complex<double>, double>::init_pot(Charge const&) /__w/abacus-develop/abacus-develop/source/module_lr/esolver_lrtd_lcao.cpp:502
    #3 0x56242300f4de in LR::ESolver_LR<std::complex<double>, double>::ESolver_LR(ModuleESolver::ESolver_KS_LCAO<std::complex<double>, double>&&, Input_para const&, UnitCell&) /__w/abacus-develop/abacus-develop/source/module_lr/esolver_lrtd_lcao.cpp:187
    #4 0x562422894968 in ModuleESolver::init_esolver(Input_para const&, UnitCell&) /__w/abacus-develop/abacus-develop/source/module_esolver/esolver.cpp:230
    #5 0x562422456845 in Driver::driver_run() /__w/abacus-develop/abacus-develop/source/driver_run.cpp:49
    #6 0x5624224503de in Driver::atomic_world() /__w/abacus-develop/abacus-develop/source/driver.cpp:188
    #7 0x5624224555d6 in Driver::init() /__w/abacus-develop/abacus-develop/source/driver.cpp:41
    #8 0x562421afe18a in main /__w/abacus-develop/abacus-develop/source/main.cpp:42

Direct leak of 30 byte(s) in 1 object(s) allocated from:
    #0 0x7f766c380887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7f76293ec4e6  (<unknown module>)

Direct leak of 23 byte(s) in 1 object(s) allocated from:
    #0 0x7f766c3279a7 in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cpp:454
    #1 0x7f762b9b7f23  (<unknown module>)

Direct leak of 20 byte(s) in 1 object(s) allocated from:
    #0 0x7f766c3279a7 in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cpp:454
    #1 0x7f762932b570  (<unknown module>)
    #2 0x3331312e32396334  (<unknown module>)

Direct leak of 5 byte(s) in 1 object(s) allocated from:
    #0 0x7f766c3279a7 in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cpp:454
    #1 0x7f762b9b7d65  (<unknown module>)

Direct leak of 1 byte(s) in 1 object(s) allocated from:
    #0 0x7f766c3279a7 in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cpp:454
    #1 0x7f76294200f4  (<unknown module>)

SUMMARY: AddressSanitizer: 14270 byte(s) leaked in 33 allocation(s).
```

### Expected behavior

_No response_

### To Reproduce

_No response_

### Environment

_No response_

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Verify the issue is not a duplicate.
- [ ] Describe the bug.
- [ ] Steps to reproduce.
- [ ] Expected behavior.
- [ ] Error message.
- [ ] Environment details.
- [ ] Additional context.
- [ ] Assign a priority level (low, medium, high, urgent).
- [ ] Assign the issue to a team member.
- [ ] Label the issue with relevant tags.
- [ ] Identify possible related issues.
- [ ] Create a unit test or automated test to reproduce the bug (if applicable).
- [ ] Fix the bug.
- [ ] Test the fix.
- [ ] Update documentation (if necessary).
- [ ] Close the issue and inform the reporter (if applicable).

### PR DESCRIPTION
**Pull request by maki49**
Fix #4718 
(I think #4718 is caused by a forgotten delete of `PotHartree`.)

### CODE REVIEW

**Review by mohanchen — APPROVED** [review]

**Comment by kirk0830** [comment]
For all use of shared_ptr, my question would be: why shared_ptr? why not unique_ptr? what special ownership determines the use of shared_ptr?

**Review by kirk0830 — CHANGES_REQUESTED** [review]
For all use of shared_ptr, my question would be: why shared_ptr? why not unique_ptr? what special ownership determines the use of shared_ptr?

**Comment by maki49** [comment]
> For all use of shared_ptr, my question would be: why shared_ptr? why not unique_ptr? what special ownership determines the use of shared_ptr?

`DensityMatrix<T,T>* DM_trans`, `HContainer<T>* hR` and `Exx_LRI<T>* exx_lri` are members shared in different classes so they should be shared_ptr.
For `PotHartree`, I can use unique_ptr, but because `std::make_unique` is missing in C++11, I just use shared_ptr for convenience. Should I change it to unique_ptr?

**Comment by kirk0830** [comment]
> > For all use of shared_ptr, my question would be: why shared_ptr? why not unique_ptr? what special ownership determines the use of shared_ptr?
> 
> `DensityMatrix<T,T>* DM_trans`, `HContainer<T>* hR` and `Exx_LRI<T>* exx_lri` are members shared in different classes so they should be shared_ptr. For `PotHartree`, I can use unique_ptr, but because `std::make_unique` is missing in C++11, I just use shared_ptr for convenience. Should I change it to unique_ptr?

Not exactly. The pointer manages memory can be shared_ptr, if "its" memory is designed to be visited by some vars other than itself. Then use weak_ptr instead of shared_ptr to get a temporary ref from the memory shared. For cases the memory is not designed to be shared, use unique_ptr.
Please see the implementation of psi_initializer, the psig is managed by shared_ptr and in esolver it is weak_ptr that visit its memory instead of another shared_ptr.
In summary, please make clear about the life span and ownership of memory, smart pointer is designed to help developer take care of practical allocate and deallocate operation but not for "smartly" manage the memory.

**Comment by maki49** [comment]
> In summary, please make clear about the life span and ownership of memory, smart pointer is designed to help developer take care of practical allocate and deallocate operation but not for "smartly" manage the memory.

In the newest commit I've changed shared_ptr to unique_ptr or weak_ptr if possible. Thanks for pointing out the problem.

**Comment by jinzx10** [comment]
I would like to put in my two cents on smart pointers.

The introduction of smart pointers in c++ is tightly connected to ownership, a concept not possessed by raw pointers. Raw pointers are mostly about access rights (as well as modification right if not pointers to const); they have no mechanism about managing the resource they point to themselves, i.e., a user has to explicitly call "delete p" to release the resource.

As one may be aware, access right is not the synonym of ownership. Changing all raw pointers to shared_ptr regardless of what the actual ownership is supposed to be might work in many situations, but doing so not only conveys less accurate information about the intention of the design, but also has the risk of introducing memory leaks that are difficult to identify. Consider the following minimal example:
```cpp
#include <memory>
#include <cstdio>

struct B; // forward declaration
struct A {
    ~A() { printf("A dtor\n"); }
    std::shared_ptr<B> ptr_;
};

struct B {
    ~B() { printf("B dtor\n"); }
    std::shared_ptr<A> ptr_;
};

int main() {

    std::shared_ptr<A> a = std::make_shared<A>();
    std::shared_ptr<B> b = std::make_shared<B>();

    a->ptr_ = b;
    b->ptr_ = a;

    return 0;
}
```
The code compiles and runs without error, but you won't see any output as supposed to be printed by the destructor, and the sanitizer would report memory leaks. What's happening here is that, before returning from main(), reference counting for both shared_ptrs are 2 (instead of 1): 1 for the object defined in main(), and extra 1 for member variables. When b goes out of scope, reference count of shared_ptr\<B\> is merely reduced by 1, so the heap-allocated B object is not destructed.

While the above circular dependency is easy to detect, similar bugs might be extremely difficult to disclose in complicated dependencies between multiple objects. The root of this kind of bugs is "abusing ownership": it hardly makes sense for two classes to "own" each other. Normally one class own the other, and the other merely needs to access its owner. The latter relation is precisely the motivation of weak_ptr: 
```cpp
#include <memory>
#include <cstdio>

struct B; // forward declaration
struct A {
    ~A() { printf("A dtor\n"); }
    std::shared_ptr<B> ptr_;  // <---- with ownership
};

struct B {
    ~B() { printf("B dtor\n"); }
    std::weak_ptr<A> ptr_; // <---- without ownership
};

int main() {

    std::shared_ptr<A> a = std::make_shared<A>();
    std::shared_ptr<B> b = std::make_shared<B>();

    a->ptr_ = b;
    b->ptr_ = a;

    return 0;
}
```
In the above example, class A is supposed to have a shared ownership of B, while class B merely has access right to A. The above code should run without memory leaks.

To summarize, I think smart pointers should not be considered ``silver bullets'' of memory issues. They are not intelligent enough to understand the purposes of developers, nor are they capable of resolving the onerous task of memory management without careful treatment. I think they're just little tools that help save some effort only <B>after</B> developers have figured out the right ownership relation. Without ownership in mind, shared_ptr could be even more risky than raw pointers.

PS: while make_unique is not available by the standard library of c++11, it can be easily implemented with c++11 syntax and is basically the following:
```cpp
template<typename T, typename... Args>
std::unique_ptr<T> make_unique(Args&&...args) {
    return unique_ptr<T>(new T(std::forward<Args>(args)...));
}
```
(here's a snapshot from /usr/include/c++/11/bits/unique_ptr.h)
image

**Comment by kirk0830** [comment]
@maki49 Hi, I have reviewed your PR and note your changes replacing shared_ptr with more appropriate choices, but could you please scan with AddressSanitizer to check if related memory leaks have been fixed? Thanks.

**Comment by maki49** [comment]
> @maki49 Hi, I have reviewed your PR and note your changes replacing shared_ptr with more appropriate choices, but could you please scan with AddressSanitizer to check if related memory leaks have been fixed? Thanks.

Seems fixed. The full message is the same as other cases:
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==3439662==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000493 (pc 0x000000000493 bp 0x7fffa8cdc105 sp 0x7fffa8cdc058 T0)
==3439662==Hint: pc points to the zero page.
==3439662==The signal is caused by a READ memory access.
==3439662==Hint: address points to the zero page.
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer: nested bug in the same thread, aborting.
AddressSanitizer:DEADLYSIGNAL
=================================================================
AddressSanitizer:DEADLYSIGNAL
=================================================================
==3439664==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000493 (pc 0x000000000493 bp 0x7ffd156c58e5 sp 0x7ffd156c5838 T0)
==3439664==Hint: pc points to the zero page.
==3439664==The signal is caused by a READ memory access.
==3439664==Hint: address points to the zero page.
AddressSanitizer:DEADLYSIGNAL
==3439663==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000493 (pc 0x000000000493 bp 0x7ffc6fd0b005 sp 0x7ffc6fd0af58 T0)
==3439663==Hint: pc points to the zero page.
==3439663==The signal is caused by a READ memory access.
=================================================================
==3439663==Hint: address points to the zero page.
==3439665==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000493 (pc 0x000000000493 bp 0x7ffce20d9d45 sp 0x7ffce20d9c98 T0)
==3439665==Hint: pc points to the zero page.
==3439665==The signal is caused by a READ memory access.
==3439665==Hint: address points to the zero page.
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer: nested bug in the same thread, aborting.
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer: nested bug in the same thread, aborting.
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer: nested bug in the same thread, aborting.
```

**Comment by kirk0830** [comment]
> > @maki49 Hi, I have reviewed your PR and note your changes replacing shared_ptr with more appropriate choices, but could you please scan with AddressSanitizer to check if related memory leaks have been fixed? Thanks.
> 
> Seems fixed. The full message is the same as other cases:
> 
> ```
> AddressSanitizer:DEADLYSIGNAL
> =================================================================
> ==3439662==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000493 (pc 0x000000000493 bp 0x7fffa8cdc105 sp 0x7fffa8cdc058 T0)
> ==3439662==Hint: pc points to the zero page.
> ==3439662==The signal is caused by a READ memory access.
> ==3439662==Hint: address points to the zero page.
> AddressSanitizer:DEADLYSIGNAL
> AddressSanitizer: nested bug in the same thread, aborting.
> AddressSanitizer:DEADLYSIGNAL
> =================================================================
> AddressSanitizer:DEADLYSIGNAL
> =================================================================
> ==3439664==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000493 (pc 0x000000000493 bp 0x7ffd156c58e5 sp 0x7ffd156c5838 T0)
> ==3439664==Hint: pc points to the zero page.
> ==3439664==The signal is caused by a READ memory access.
> ==3439664==Hint: address points to the zero page.
> AddressSanitizer:DEADLYSIGNAL
> ==3439663==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000493 (pc 0x000000000493 bp 0x7ffc6fd0b005 sp 0x7ffc6fd0af58 T0)
> ==3439663==Hint: pc points to the zero page.
> ==3439663==The signal is caused by a READ memory access.
> =================================================================
> ==3439663==Hint: address points to the zero page.
> ==3439665==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000493 (pc 0x000000000493 bp 0x7ffce20d9d45 sp 0x7ffce20d9c98 T0)
> ==3439665==Hint: pc points to the zero page.
> ==3439665==The signal is caused by a READ memory access.
> ==3439665==Hint: address points to the zero page.
> AddressSanitizer:DEADLYSIGNAL
> AddressSanitizer: nested bug in the same thread, aborting.
> AddressSanitizer:DEADLYSIGNAL
> AddressSanitizer: nested bug in the same thread, aborting.
> AddressSanitizer:DEADLYSIGNAL
> AddressSanitizer: nested bug in the same thread, aborting.
> ```

@caic99 Hi, could you help see if the output means there is no direct memory leak anymore? thanks

**Comment by caic99** [comment]
> @caic99 Hi, could you help see if the output means there is no direct memory leak anymore? thanks

@kirk0830 Based on this excerpt I'm not able to determine if it happens anymore. It seems that the program halted somewhere inside the execution process.
However, I suggest we merging this PR first, and then run dynamic analysis again to determine the status.

**Review by kirk0830 — APPROVED** [review]
